### ISSUE TITLE
Mark certain exceptions as handled in onError()

### ISSUE THREAD

**Issue by ayushin** [felangel/bloc#1254]
Just looking at https://github.com/felangel/bloc/issues/1122

What about the following design pattern:

I am handling the API errors throughout the bloc's in the BlocDelegate.onError() method by showing a message to the user to avoid endless try... catch boilerplate in the blocs.

A lot of api blocs error handling is as simple as showing the error message and stop at the point of the exception, and if something extra is needed it can be handled in the bloc's onError.

It would be nice if I could mark those exceptions as handled to avoid unnecessary `unhandled exceptions` messages in the debug mode.

Would a simple `ignore: true` extra argument to the super onError work for this case?

example, what i'd like to be able to do in the delegate:

```
  @override
  void onError(Bloc bloc, Object error, StackTrace stacktrace) {
    messageBloc?.add(MessageError(exception: error));
    super.onError(bloc, error, stacktrace, ignore: true);
  }

```

Well in this case I actually don't really want to call `super.onError` but as it is no longer allowed, what would be a better solution if not the above?

**Comment by felangel** [comment]
Hi @ayushin 👋 
Thanks for opening an issue!

I don't think you want to avoid calling `super.onError` because then the `BlocDelegate` will not be notified so your `messageBloc` will never receive the event. Also, the unhandled exception occurs within the bloc rather than the `BlocDelegate` so I don't think that's a viable solution to your issue.

Can you please elaborate on what the specific issue is? If I understand correctly, you just don't want to see the unhandled exception printed in the debug console? I suppose we can make that configurable but I feel as though it's useful to see that information as a developer and it doesn't negatively impact the remaining application flow so I would recommend leaving this behavior as is. Let me know what you think or if I misunderstood the issue and thanks again 👍

**Comment by ayushin** [comment]
Hi @felangel! Thanks for the quick response and also wanted to let you know you're doing a terrific job on the bloc!

The way I'd like to handle exceptions, API exceptions in particular is using the delegate's `onErorr` as a single point for displaying the error message coming from the backend to the user and I'm also using the specific bloc onError() override to yield a block's error state.

This keeps the actual mapToEvent very lean avoiding the need to wrap the API calls in the boiler plate try ... catch in every bloc  around every block fo api calls.

So if that's an API error I handle this in onError and after that I would like to avoid UnhandledException re-throwing because in fact those exceptions are handled in the onError (for example 401 Authentication Error when a user types wrong credentials) and it would make the real unhandled exceptions harder to find in the logs.

The super.onErorr call I wanted to skip would be not in the bloc, but inside BlogDelegate's onError already.

So basically I want the final assert'ed `throw BlocUnhandledErrorException` to be suppressed for the exceptions that were already handled in the bloc delegate.

What about making `BlocSupervisor.delegate.onError` returning an exception if it is unhandled or null if there's no need to further throw an exceptinos,  something like:
```
  @mustCallSuper
  void onError(Object error, StackTrace stackTrace) {
    if(BlocSupervisor.delegate.onError(this, error, stackTrace) != null) {
    assert(() { 
      throw BlocUnhandledErrorException(this, error, stackTrace);
    }());
   }
  }
```
and make the default to be

`BlocSupervisor.delegate.onError(this, error, stackTrace) => error`

Then returning null value would allow marking exception as handled.

**Comment by ayushin** [comment]
I made a non-breaking PR to illustrate my point:

https://github.com/felangel/bloc/pull/1265

**Comment by ayushin** [comment]
I'd say we don't even really need to have BlocUnhandledErrorException at all, it would suffice to replace it with:

```
    assert(BlocSupervisor.delegate.onError(this, error, stackTrace) != null);
```

and require BlocSupervisor.delegate.onError() to return a non-null value if the exception is handled.

**Comment by felangel** [comment]
Closing this for the reasons described in https://github.com/felangel/bloc/pull/1357#issuecomment-651856042.

Let me know if you have additional concerns or questions and I'm happy to continue the conversation.
TL;DR I don't think BlocDelegate/BlocObserver is the right place for global error handling. I think it should only be used as a passive element to inspect/track that information but not to intervene. In your case, I would highly recommend restructuring your application architecture to have the errors be handled at the repository layer in a unified way. Then you can have a global error handling bloc which consumes the errors from all repositories and triggers an overlay or dialog.

Hope that helps 👍
